parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

```diff
@@ -192,324 +190,351 @@ sqliteCodebase ::
+
+            namesAtPath :: Path -> m ScopedNames
+            namesAtPath path =
+              runTransaction (CodebaseOps.namesAtPath path)
+
+            updateNameLookup :: m ()
+            updateNameLookup =
+              runTransaction (CodebaseOps.updateNameLookupIndexFromV2Root getDeclType)
+
+        let codebase =
+              C.Codebase
+                { getTerm = Cache.applyDefined termCache getTerm,
+                  getTypeOfTermImpl = Cache.applyDefined typeOfTermCache getTypeOfTermImpl,
+                  getTypeDeclaration = Cache.applyDefined declCache getTypeDeclaration,
+                  getDeclType =
+                    \r ->
+                      withConn \conn ->
+                        Sqlite.runReadOnlyTransaction conn \run -> run (getDeclType r),
+                  putTerm,
+                  putTypeDeclaration,
+                  getTermComponentWithTypes,
+                  getDeclComponent,
+                  getComponentLength = getCycleLength,
+                  getRootBranch = getRootBranch rootBranchCache,
+                  getRootBranchHash,
+                  getRootBranchExists,
+                  putRootBranch = putRootBranch rootBranchCache,
+                  rootBranchUpdates = rootBranchUpdates rootBranchCache,
+                  getShallowBranchForHash,
+                  getBranchForHashImpl = getBranchForHash,
+                  putBranch,
+                  branchExists = isCausalHash,
+                  getPatch,
+                  putPatch,
+                  patchExists,
+                  dependentsImpl,
+                  dependentsOfComponentImpl,
+                  syncFromDirectory,
+                  syncToDirectory,
+                  viewRemoteBranch',
+                  pushGitBranch = \repo opts action -> withConn \conn -> pushGitBranch conn repo opts action,
+                  watches,
+                  getWatch,
+                  putWatch,
+                  clearWatches,
+                  getReflog,
+                  appendReflog,
+                  termsOfTypeImpl,
+                  termsMentioningTypeImpl,
+                  hashLength,
+                  termReferencesByPrefix,
+                  typeReferencesByPrefix = declReferencesByPrefix,
+                  termReferentsByPrefix = referentsByPrefix,
+                  branchHashLength,
+                  branchHashesByPrefix,
+                  lcaImpl = Just sqlLca,
+                  beforeImpl,
+                  namesAtPath,
+                  updateNameLookup,
+                  withConnection = withConn,
+                  withConnectionIO = withConnection debugName root
+                }
+        Right <$> action codebase
+  where
+    withConn :: (Sqlite.Connection -> m a) -> m a
+    withConn =
+      withConnection debugName root
+
+    runTransaction :: Sqlite.Transaction a -> m a
+    runTransaction action =
+      withConn \conn -> Sqlite.runTransaction conn action
```

Comment:
I'm sure it's fine, but is it possible that using a single connection shared across the codebase gave us any benefits in terms of in-memory caching?

If there's anywhere we're querying the codebase in a tight-loop (there's probably at least some spots that do this) we're now acquiring a new connection for each and every query rather than using a single one.

I have no idea whether this is likely to actually cause problems in reality though 🤷🏼‍♂️

Reply:
I do think it'd be wiser to reuse connections wherever possible - @tstat suggested we just use a connection pool. In the meantime, I'm not sure about where we might be hitting these functions in a tight loop - I did time loading the root branch into memory, and there wasn't any noticeable slowdown. Push/pull to Share, at least, aren't using this `Codebase` interface for storing terms, so we don't have a blowup of connection creation/destruction over there.

Reply:
Yeah, I thought about a pool but wasn't sure whether it makes sense or not given how much lighter-weight sqlite connections are. Wouldn't hurt though, and probably would be easy to wire into what you've got here right?

If it's not showing any noticeable difference maybe it's fine?